BUG: assert_series_equal(..., check_exact=True, check_dtype=False) may still raise if dtypes don't match

### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.


### Reproducible Example

```python
left = Series([10], dtype='Int64')
right = Series([10], dtype='int64')
pd.testing.assert_series_equal(left, right, check_exact=True, check_dtype=False)
```


### Issue Description

this raises
```python
AssertionError                            Traceback (most recent call last)
Cell In[26], line 3
      1 left = Series([10], dtype='Int64')
      2 right = Series([10], dtype='int64')
----> 3 pd.testing.assert_series_equal(left, right, check_exact=True, check_dtype=False)

    [... skipping hidden 3 frame]

File ~/tmp/.venv/lib/python3.10/site-packages/pandas/_testing/asserters.py:611, in raise_assert_detail(obj, message, left, right, diff, first_diff, index_values)
    608 if first_diff is not None:
    609     msg += f"\n{first_diff}"
--> 611 raise AssertionError(msg)

AssertionError: Series are different

Series classes are different
[left]:  IntegerArray
[right]: ndarray
```

### Expected Behavior

this shouldn't raise

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : 1c606d5f014c5296d6028af28001311b67ee3721
python                : 3.10.12.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 5.10.102.1-microsoft-standard-WSL2
Version               : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : en_GB.UTF-8
LOCALE                : en_GB.UTF-8

pandas                : 2.2.0.dev0+694.g1c606d5f01
numpy                 : 1.25.1
pytz                  : 2023.3
dateutil              : 2.8.2
setuptools            : 67.6.1
pip                   : 23.1.2
Cython                : None
pytest                : 7.3.1
hypothesis            : 6.82.4
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.2
IPython               : 8.16.1
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.2
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : 2023.7.0
fsspec                : 2023.6.0
gcsfs                 : None
matplotlib            : 3.7.1
numba                 : 0.58.1
numexpr               : None
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : 12.0.1
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.11.0
sqlalchemy            : 2.0.20
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.3
qtpy                  : None
pyqt5                 : None

</details>

--- a/pandas/_testing/asserters.py
+++ b/pandas/_testing/asserters.py
@@ -949,9 +949,15 @@ def assert_series_equal(
                 obj=str(obj),
             )
         else:
+            # convert both to NumPy if not, check_dtype would raise earlier
+            lv, rv = left_values, right_values
+            if isinstance(left_values, ExtensionArray):
+                lv = left_values.to_numpy()
+            if isinstance(right_values, ExtensionArray):
+                rv = right_values.to_numpy()
             assert_numpy_array_equal(
-                left_values,
-                right_values,
+                lv,
+                rv,
                 check_dtype=check_dtype,
                 obj=str(obj),
                 index_values=left.index,

--- a/pandas/tests/extension/test_numpy.py
+++ b/pandas/tests/extension/test_numpy.py
@@ -421,10 +421,6 @@ class TestNumpyExtensionArray(base.ExtensionTests):
     def test_EA_types(self, engine, data, request):
         super().test_EA_types(engine, data, request)
 
-    @pytest.mark.xfail(reason="Expect NumpyEA, get np.ndarray")
-    def test_compare_array(self, data, comparison_op):
-        super().test_compare_array(data, comparison_op)
-
     def test_compare_scalar(self, data, comparison_op, request):
         if data.dtype.kind == "f" or comparison_op.__name__ in ["eq", "ne"]:
             mark = pytest.mark.xfail(reason="Expect NumpyEA, get np.ndarray")

--- a/pandas/tests/extension/test_numpy.py
+++ b/pandas/tests/extension/test_numpy.py
@@ -421,12 +421,6 @@ class TestNumpyExtensionArray(base.ExtensionTests):
     def test_EA_types(self, engine, data, request):
         super().test_EA_types(engine, data, request)
 
-    def test_compare_scalar(self, data, comparison_op, request):
-        if data.dtype.kind == "f" or comparison_op.__name__ in ["eq", "ne"]:
-            mark = pytest.mark.xfail(reason="Expect NumpyEA, get np.ndarray")
-            request.applymarker(mark)
-        super().test_compare_scalar(data, comparison_op)
-
 
 class Test2DCompat(base.NDArrayBacked2DTests):
     pass

--- a/pandas/tests/util/test_assert_frame_equal.py
+++ b/pandas/tests/util/test_assert_frame_equal.py
@@ -211,10 +211,7 @@ def test_assert_frame_equal_extension_dtype_mismatch():
         "\\[right\\]: int[32|64]"
     )
 
-    # TODO: this shouldn't raise (or should raise a better error message)
-    # https://github.com/pandas-dev/pandas/issues/56131
-    with pytest.raises(AssertionError, match="classes are different"):
-        tm.assert_frame_equal(left, right, check_dtype=False)
+    tm.assert_frame_equal(left, right, check_dtype=False)
 
     with pytest.raises(AssertionError, match=msg):
         tm.assert_frame_equal(left, right, check_dtype=True)
@@ -246,7 +243,6 @@ def test_assert_frame_equal_ignore_extension_dtype_mismatch():
     tm.assert_frame_equal(left, right, check_dtype=False)
 
 
-@pytest.mark.xfail(reason="https://github.com/pandas-dev/pandas/issues/56131")
 def test_assert_frame_equal_ignore_extension_dtype_mismatch_cross_class():
     # https://github.com/pandas-dev/pandas/issues/35715
     left = DataFrame({"a": [1, 2, 3]}, dtype="Int64")
@@ -300,9 +296,7 @@ def test_frame_equal_mixed_dtypes(frame_or_series, any_numeric_ea_dtype, indexer
     dtypes = (any_numeric_ea_dtype, "int64")
     obj1 = frame_or_series([1, 2], dtype=dtypes[indexer[0]])
     obj2 = frame_or_series([1, 2], dtype=dtypes[indexer[1]])
-    msg = r'(Series|DataFrame.iloc\[:, 0\] \(column name="0"\) classes) are different'
-    with pytest.raises(AssertionError, match=msg):
-        tm.assert_equal(obj1, obj2, check_exact=True, check_dtype=False)
+    tm.assert_equal(obj1, obj2, check_exact=True, check_dtype=False)
 
 
 def test_assert_frame_equal_check_like_different_indexes():

--- a/pandas/tests/util/test_assert_series_equal.py
+++ b/pandas/tests/util/test_assert_series_equal.py
@@ -290,10 +290,7 @@ Attribute "dtype" are different
 \\[left\\]:  Int64
 \\[right\\]: int[32|64]"""
 
-    # TODO: this shouldn't raise (or should raise a better error message)
-    # https://github.com/pandas-dev/pandas/issues/56131
-    with pytest.raises(AssertionError, match="Series classes are different"):
-        tm.assert_series_equal(left, right, check_dtype=False)
+    tm.assert_series_equal(left, right, check_dtype=False)
 
     with pytest.raises(AssertionError, match=msg):
         tm.assert_series_equal(left, right, check_dtype=True)
@@ -372,7 +369,6 @@ def test_assert_series_equal_ignore_extension_dtype_mismatch():
     tm.assert_series_equal(left, right, check_dtype=False)
 
 
-@pytest.mark.xfail(reason="https://github.com/pandas-dev/pandas/issues/56131")
 def test_assert_series_equal_ignore_extension_dtype_mismatch_cross_class():
     # https://github.com/pandas-dev/pandas/issues/35715
     left = Series([1, 2, 3], dtype="Int64")

--- a/pandas/tests/util/test_assert_series_equal.py
+++ b/pandas/tests/util/test_assert_series_equal.py
@@ -456,3 +456,12 @@ def test_large_unequal_ints(dtype):
     right = Series([1577840521123543], dtype=dtype)
     with pytest.raises(AssertionError, match="Series are different"):
         tm.assert_series_equal(left, right)
+
+
+@pytest.mark.parametrize("check_exact", [True, False])
+@pytest.mark.parametrize("val", [3, 3.5])
+def test_ea_and_numpy_no_dtype_check(val, check_exact):
+    # GH#56651
+    left = Series([1, 2, val])
+    right = Series(pd.array([1, 2, val]))
+    tm.assert_series_equal(left, right, check_dtype=False, check_exact=check_exact)

--- a/pandas/tests/util/test_assert_series_equal.py
+++ b/pandas/tests/util/test_assert_series_equal.py
@@ -454,10 +454,11 @@ def test_large_unequal_ints(dtype):
         tm.assert_series_equal(left, right)
 
 
+@pytest.mark.parametrize("dtype", [None, object])
 @pytest.mark.parametrize("check_exact", [True, False])
 @pytest.mark.parametrize("val", [3, 3.5])
-def test_ea_and_numpy_no_dtype_check(val, check_exact):
+def test_ea_and_numpy_no_dtype_check(val, check_exact, dtype):
     # GH#56651
-    left = Series([1, 2, val])
+    left = Series([1, 2, val], dtype=dtype)
     right = Series(pd.array([1, 2, val]))
     tm.assert_series_equal(left, right, check_dtype=False, check_exact=check_exact)

